tests/infra/logging_app.py

```diff
@@ -4,82 +4,163 @@
 import infra.checker
 import time
 import http
+import ccf.clients
+from collections import defaultdict
 
 from loguru import logger as LOG
 
 
 class LoggingTxs:
     def __init__(self, user_id=0):
-        self.pub = {}
-        self.priv = {}
-        self.next_pub_index = 1
-        self.next_priv_index = 1
+        self.pub = defaultdict(list)
+        self.priv = defaultdict(list)
+        self.idx = 0
         self.user = f"user{user_id}"
 
     def issue(
         self,
         network,
-        number_txs,
+        number_txs=1,
         on_backup=False,
+        repeat=False,
     ):
         remote_node, _ = network.find_primary()
         if on_backup:
             remote_node = network.find_any_backup()
 
         LOG.info(f"Applying {number_txs} logging txs to node {remote_node.node_id}")
 
-        with remote_node.client(self.user) as uc:
-            check_commit = infra.checker.Checker(uc)
+        with remote_node.client(self.user) as c:
+            check_commit = infra.checker.Checker(c)
+
             for _ in range(number_txs):
-                priv_msg = f"Private message at index {self.next_priv_index}"
-                pub_msg = f"Public message at index {self.next_pub_index}"
-                rep_priv = uc.post(
+                if not repeat:
+                    self.idx += 1
+
+                priv_msg = (
+                    f"Private message at idx {self.idx} [{len(self.priv[self.idx])}]"
+                )
+                rep_priv = c.post(
                     "/app/log/private",
                     {
-                        "id": self.next_priv_index,
+                        "id": self.idx,
                         "msg": priv_msg,
                     },
                 )
-                rep_pub = uc.post(
-                    "/app/log/public",
-                    {
-                        "id": self.next_pub_index,
-                        "msg": pub_msg,
-                    },
-                )
                 check_commit(rep_priv, result=True)
-                check_commit(rep_pub, result=True)
+                self.priv[self.idx].append(
+                    {"msg": priv_msg, "seqno": rep_priv.seqno, "view": rep_priv.view}
+                )
 
-                self.priv[self.next_priv_index] = priv_msg
-                self.pub[self.next_pub_index] = pub_msg
-                self.next_priv_index += 1
-                self.next_pub_index += 1
+                # Public records do not handle historical queries
```

Comment:
We don't have an endpoint to query the public table in the logging app, but there's nothing preventing us from adding it. I think this client code should do the same thing for writing to public and private tables (ie - remove the `if not repeat`), and simply skip the `/historical` calls when verifying the public entries. In particular, this level of repeat tests something that we're otherwise _not_ testing, that entries are overwritten when we reuse IDs.

Reply:
Good point. I didn't want to add a historical endpoint for public records as well as this feature is still in progress and the historical endpoint is fairly verbose for now. I've updated the `logging_app.py` though so that public records are repeated as well but only the latest submitted entry is verified. We can revisit this later when the support for historical queries is complete.